apps/checkout-app/pages/api/webhooks/adyen.ts

```diff
@@ -4,38 +4,73 @@ import { NextApiRequest, NextApiResponse } from "next";
 import { hmacValidator, Types } from "@adyen/api-library";
 
 import { createTransaction } from "@/backend/payments/createTransaction";
-import { verifyPayment } from "@/backend/payments/providers/adyen";
+import {
+  getNewTransactionData,
+  getOrderId,
+  getUpdatedTransactionData,
+  isNotificationDuplicate,
+} from "@/backend/payments/providers/adyen";
+import { getOrderTransactions } from "@/backend/payments/getOrderTransactions";
+import { updateTransaction } from "@/backend/payments/updateTransaction";
 
 const HMAC = process.env.ADYEN_HMAC!;
 
 const validator = new hmacValidator();
 
-const validateNotificationItems = (
-  notificationItems: Types.notification.NotificationItem[]
-) =>
-  notificationItems.map(
-    ({ NotificationRequestItem }: Types.notification.NotificationItem) => {
-      // first validate the origin
-      const valid = validator.validateHMAC(NotificationRequestItem, HMAC);
+const validateNotificationItems = ({
+  NotificationRequestItem,
+}: Types.notification.NotificationItem) => {
+  // first validate the origin
+  const valid = validator.validateHMAC(NotificationRequestItem, HMAC);
 
-      if (!valid) {
-        throw "Invalid HMAC key";
-      }
+  if (!valid) {
+    throw "Invalid HMAC key";
+  }
 
-      return NotificationRequestItem;
-    }
-  );
+  return NotificationRequestItem;
+};
 
 const notificationHandler = async (
   notification: Types.notification.NotificationRequestItem
 ) => {
-  const data = await verifyPayment(notification);
+  // Get order id from webhook metadata
+  const orderId = await getOrderId(notification);
+
+  if (!orderId) {
+    return;
+  }
 
-  if (!data) {
+  // Get order transactions and run deduplication
+  // https://docs.adyen.com/development-resources/webhooks/best-practices#handling-duplicates
+  const transactions = await getOrderTransactions({ id: orderId });
+  const duplicate = await isNotificationDuplicate(transactions, notification);
+
+  if (duplicate) {
     return;
   }
 
-  createTransaction(data);
+  // Check if originalReference exists, if it does append a new event
+  if (notification.originalReference) {
+    const transaction = transactions.find(
+      ({ reference }) => reference === notification.originalReference
+    );
+
+    if (!transaction) {
+      throw "originalReference does not exist in transactions";
+    }
+
+    const data = await getUpdatedTransactionData(transaction, notification);
+
+    updateTransaction(data);
+  } else {
+    const data = await getNewTransactionData(orderId, notification);
+
+    if (!data) {
```

Comment:
What does this case mean? It doesn't throw an error but doesn't create a transaction either

Reply:
There's an error handling inside, basically we skip webhook events that we don't handle at the moment